### PR TITLE
[feat] Add Average Endpoint Error to Metrics

### PR DESCRIPTION
**Pull request by ChristophReich1996**
#### Changes
This pull request implements the average endpoint error (AEPE) commonly used to evaluate optical flow estimation approaches (e.g., [PWC-Net](https://arxiv.org/abs/1709.02371)).

#### Type of change

- [X] 🧪 Tests Cases
- [X] 🔬 New feature (non-breaking change which adds functionality)

#### Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/metrics/endpoint_error.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    """
+    KORNIA_CHECK_IS_TENSOR(input)
+    KORNIA_CHECK_IS_TENSOR(target)
+    KORNIA_CHECK_SHAPE(input, ["*", "2"])
+    KORNIA_CHECK_SHAPE(target, ["*", "2"])
+    KORNIA_CHECK(
+        input.shape == target.shape, f"input and target shapes must be the same. Got: {input.shape} and {target.shape}"
+    )
+
+    epe: Tensor = ((input[..., 0] - target[..., 0]) ** 2 + (input[..., 1] - target[..., 1]) ** 2).sqrt().mean()
```
Comment (edgarriba) [inline kornia/metrics/endpoint_error.py:40]: Should we add a `reduction` parameter to with options `none` , `sum` , `mean` ?

Comment (ChristophReich1996) [reply kornia/metrics/endpoint_error.py:44]: Sorry for my late response, I just added a reduction parameter to support the mentioned reduction approaches.

File: kornia/metrics/endpoint_error.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
+import torch
+from torch import Tensor, nn
+
+from kornia.core.check import KORNIA_CHECK, KORNIA_CHECK_IS_TENSOR, KORNIA_CHECK_SHAPE
+
+
+def aepe(input: torch.Tensor, target: torch.Tensor, reduction: str = "mean") -> torch.Tensor:
+    r"""Create a function that calculates the average endpoint error (AEPE) between 2 flow maps.
+
+    AEPE is the endpoint error between two 2D vecotrs (e.g., oprical flow).
```
Comment (edgarriba) [inline kornia/metrics/endpoint_error.py:10]:
```suggestion
    AEPE is the endpoint error between two 2D vectors (e.g., optical flow).
```

File: kornia/metrics/endpoint_error.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
+import torch
+from torch import Tensor, nn
+
+from kornia.core.check import KORNIA_CHECK, KORNIA_CHECK_IS_TENSOR, KORNIA_CHECK_SHAPE
+
+
+def aepe(input: torch.Tensor, target: torch.Tensor, reduction: str = "mean") -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/metrics/endpoint_error.py:7]: i'd also add an alias `average_enpoint_error = aepe` (or viceversa) to have more explicit name functions

File: kornia/metrics/endpoint_error.py (hunk 1 above)
Comment (edgarriba) [reply kornia/metrics/endpoint_error.py:44]: don't you need to remove the `mean()` call above now ?

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: kornia/metrics/endpoint_error.py (hunk 3 above)
Comment (johnnv1) [reply kornia/metrics/endpoint_error.py:7]: +1

File: test/test_metrics.py [hunk 4]
```diff
@@ -196,6 +196,21 @@ def test_metric(self, device, dtype):
         assert_close(actual, expected)
 
 
+class TestEpe:
```
Comment (johnnv1) [inline test/test_metrics.py:199]: can you also add tests for each reduction?

**Comment by ChristophReich1996** [comment]
Thanks for the feedback, I added the metric to the docs and wrote additional test cases for different reductions (etc.).

File: kornia/metrics/endpoint_error.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    else:
+        raise NotImplementedError('Invalid reduction option.')
+
+    return epe
+
+
+class AEPE(nn.Module):
+    r"""Computes the average endpoint error (AEPE) between 2 flow maps.
+
+    EPE is the endpoint error between two 2D vecotrs (e.g., oprical flow).
```
Comment (johnnv1) [inline kornia/metrics/endpoint_error.py:61]:
```suggestion
    EPE is the endpoint error between two 2D vectors (e.g., oprical flow).
```

File: kornia/metrics/endpoint_error.py (hunk 3 above)
Comment (johnnv1) [reply kornia/metrics/endpoint_error.py:7]: @ChristophReich1996 can you add this alias?

File: test/test_metrics.py [hunk 6]
```diff
@@ -196,6 +196,47 @@ def test_metric(self, device, dtype):
...
+        assert_close(actual, expected)
+
+    def test_metric_no_reduction(self, device, dtype):
+        sample = torch.ones(4, 4, 2, device=device, dtype=dtype)
+        expected = torch.zeros(4, 4, device=device, dtype=dtype) + 1.4142
+        actual = kornia.metrics.aepe(sample, 2.0 * sample, reduction="none")
+        assert_close(actual, expected)
+
+    def test_metric_wrong_reduction(self, device, dtype):
+        with pytest.raises(NotImplementedError):
```
Comment (johnnv1) [inline test/test_metrics.py:219]: you didn't need to provide test for NotImplemented raises, we consider it as a "defensive code"

File: test/test_metrics.py [hunk 7]
```diff
@@ -196,6 +196,47 @@ def test_metric(self, device, dtype):
...
+            expected = torch.zeros(4, 4, device=device, dtype=dtype) + 1.4142
+            actual = kornia.metrics.aepe(sample, 2.0 * sample, reduction="foo")
+            assert_close(actual, expected)
+
+    def test_metric_wrong_shape(self, device, dtype):
+        with pytest.raises(Exception):
+            sample = torch.ones(4, 4, 2, device=device, dtype=dtype)
+            expected = torch.zeros(4, 4, device=device, dtype=dtype) + 1.4142
+            actual = kornia.metrics.aepe(sample, 2.0 * sample[..., 0], reduction="mean")
+            assert_close(actual, expected)
```
Comment (johnnv1) [inline test/test_metrics.py:230]: Probably your code raises on calling the `aepe` itself, so the subsequent lines are not necessary
```suggestion
            kornia.metrics.aepe(sample, 2.0 * sample[..., 0], reduction="mean")
```

you can add tests for the exceptions under KORNIA_CHECK API -- you can add all into a unique test and assert the error message like here https://github.com/kornia/kornia/blob/master/test/losses/test_losses.py#L187

Comment (ChristophReich1996) [reply test/test_metrics.py]: Thanks for pointing this out. I combined all exception test into a single unique test.

File: kornia/metrics/endpoint_error.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+         ``'mean'``: the sum of the output will be divided by the number of elements
+         in the output, ``'sum'``: the output will be summed.
+
+    Return:
+        the computed AEPE as a scalar.
+
+    Examples:
+        >>> ones = torch.ones(4, 4, 2)
+        >>> aepe(ones, 1.2 * ones)
+        tensor(0.0800)
```
Comment (johnnv1) [inline kornia/metrics/endpoint_error.py:31]: can you check the failing doc string tests? 
https://github.com/kornia/kornia/pull/2500/checks#step:5:585

Comment (ChristophReich1996) [reply kornia/metrics/endpoint_error.py]: Fixed :)

**Review by johnnv1 — APPROVED** [review]
LGTM

**Review by edgarriba — APPROVED** [review]
